patroni/ha.py

```diff
@@ -447,27 +447,33 @@ def process_sync_replication(self):
         promoting standbys that were guaranteed to be replicating synchronously.
         """
         if self.is_synchronous_mode():
-            current = self.cluster.sync.leader and self.cluster.sync.sync_standby
-            picked, allow_promote = self.state_handler.pick_synchronous_standby(self.cluster)
-            if picked != current:
-                # We need to revoke privilege from current before replacing it in the config
-                if current:
-                    logger.info("Removing synchronous privilege from %s", current)
-                    if not self.dcs.write_sync_state(self.state_handler.name, None, index=self.cluster.sync.index):
+            current = self.cluster.sync.leader and self.cluster.sync.members or []
+            picked, allow_promote = self.state_handler.pick_synchronous_standby(self.cluster, sync_node_count)
+            if set(picked) != set(current):
+                # update synchronous standby list in dcs temporarily to point to common nodes in current and picked
+                sync_common = list(set(current).intersection(set(allow_promote)))
+                if set(sync_common) != set(current):
+                    logger.info("Updating synchronous privilege temporarily from %s to %s", current, sync_common)
+                    if not self.dcs.write_sync_state(self.state_handler.name,
+                                                     sync_common or None,
+                                                     index=self.cluster.sync.index):
                         logger.info('Synchronous replication key updated by someone else.')
                         return
 
-                if self.is_synchronous_mode_strict() and picked is None:
-                    picked = '*'
+                # Update  db param and wait for x secs
+                if self.is_synchronous_mode_strict() and not picked:
+                    picked = ['*']
                     logger.warning("No standbys available!")
 
                 logger.info("Assigning synchronous standby status to %s", picked)
                 self.state_handler.config.set_synchronous_standby(picked)
 
-                if picked and picked != '*' and not allow_promote:
-                    # Wait for PostgreSQL to enable synchronous mode and see if we can immediately set sync_standby
+                if picked and picked[0] != '*' and set(allow_promote) != set(picked):
+                    # Wait for PostgreSQL to enable synchronous mode and see if we can immediately
+                    # set sync_standby
                     time.sleep(2)
-                    picked, allow_promote = self.state_handler.pick_synchronous_standby(self.cluster)
+                    _, allow_promote = self.state_handler.pick_synchronous_standby(self.cluster,
+                                                                                   sync_node_count)
```

Comment:
I am thinking about little optimization here. As long as we have at least one active sync node we probably don't need to time.sleep(2), but just let it resolve on the next HA loop.

Reply:
It makes sense.  fixing the code.